proto/pulumi/resource.proto

```diff
@@ -238,3 +240,9 @@ message TransformResponse {
     google.protobuf.Struct properties = 1; // the transformed input properties.
     TransformResourceOptions options = 2; // the options for the resource.
 }
+
+message RegisterDefaultProviderRequest {
+    string provider = 1; // the provider to register as the default provider.
+    string version = 2;  // the version of the provider to register as the default provider.
```

Comment:
I don't think this is needed? The engine can work everything out from the provider instance it has in memory.

Reply:
There's a couple of problems here that I was trying to solve with adding the `version` and `pluginDownloadURL` fields:
- I don't think the provider in the engine saves the `pluginDownloadURL` anywhere, so we'll have to pass that along, as it's part of the defaultprovider map key
- Once the provider is constructed in the engine, it will always have a version.  So there's no way to tell whether it was constructed with or without a version in the SDK.  Now this is very closely related to the TODO Eron also commented on https://github.com/pulumi/pulumi/pull/16105#discussion_r1588208151.  If we decide that we should at least also always set the default provider for where the key is just `pkg`, I don't think we need to send the version along here, as we can just set `d.providers[pkg]` and `d.providers[pkg-<version>]`.  Without sending the version field along here we would only set `d.providers[pkg-<version>]`.

Reply:
> I don't think the provider in the engine saves the pluginDownloadURL anywhere, so we'll have to pass that along, as it's part of the defaultprovider map key

It should be on the provider input state.

> If we decide that we should at least also always set the default provider for where the key is just pkg

We shouldn't! But if you get given a "ref" that points to a specific provider which will have a specific version (again on its input state). We're just going to be using that provider instance as the default for it's matching pkg-version.

Reply:
Right, we'll need  #16109 for that, but then I think it works.